File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/smartcollection/SmartCollectionEntity.kt
/*
 * SmartCollectionEntity.kt
 *
 * Room @Entity for the `smart_collections` table — rule-based
 * saved views (e.g. "Invoices this month"). The rule lives in
 * `ruleJson` as an AND-of-clauses array; see brief §3 for the v1
 * grammar.
 *
 * Schema mirrors
 * `shared/design-system/migrations/quickink/v4_workspace.sql`.
 */

package app.quickink.mobile.data.smartcollection

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "smart_collections",
    indices = [
        Index(value = ["user_id", "position"], name = "idx_smart_collections_user_position"),
        Index(value = ["dirty"], name = "idx_smart_collections_dirty"),
        Index(value = ["deleted_at"], name = "idx_smart_collections_tombstone"),
    ],
)
data class SmartCollectionEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "user_id")
    val userId: String,

    @ColumnInfo(name = "name")
    val name: String,

    /**
     * Tabler icon name, e.g. "ti-receipt", "ti-eye",
     * "ti-signature". Stored as a string so the icon palette can
     * grow without a migration. NULL → default sparkle icon.
     */
    @ColumnInfo(name = "icon")
    val icon: String? = null,

    /**
     * Hex color for the icon background tint in the home strip
     * card. NULL → accent-tint default.
     */
    @ColumnInfo(name = "color")
    val color: String? = null,

    /**
     * AND-of-clauses array as canonical JSON text. See brief §3
     * for the v1 grammar — six clause types, no OR, no nesting.
     * Validated in app code (the [ruleJson] string round-trips
     * through the canonical-JSON serializer), not by the DB.
     *
     * Example seeded value for "Invoices this month":
     * `[{"type":"folder_is","folder_id":"<id>"},
     *   {"type":"date_range","field":"created_at","preset":"this_month"}]`
     */
    @ColumnInfo(name = "rule_json")
    val ruleJson: String,

    @ColumnInfo(name = "position", defaultValue = "0")
    val position: Int = 0,

    /**
     * True for shipped seed collections ("Invoices this month",
     * "Needs review", "Contains signatures"). Lets us silently
     * update the seeded ruleJson in a later release without
     * overwriting user-created collections.
     */
    @ColumnInfo(name = "is_seeded", defaultValue = "0")
    val isSeeded: Boolean = false,

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String? = null,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,
)

File: apps/quickink/ios/QuickInk/Data/SmartCollectionEntity.swift
/*
 * SmartCollectionEntity.swift
 *
 * GRDB record for the `smart_collections` table — rule-based
 * saved views (e.g. "Needs review"). The rule lives in
 * `ruleJson` as an AND-of-clauses array (see
 * `SmartCollectionRule.swift` for the v1 grammar).
 *
 * Schema mirrors
 * `shared/design-system/migrations/quickink/v4_workspace.sql`
 * + the iOS v8_workspace migration in `QuickInkDatabase.swift`.
 *
 * Mirror of `SmartCollectionEntity.kt` in QuickInk's Android
 * target.
 */

import Foundation
import GRDB

public struct SmartCollectionEntity: Codable, FetchableRecord, PersistableRecord, Equatable, Sendable, Identifiable {

    public static let databaseTableName = "smart_collections"

    public var id: String
    public var userId: String
    public var name: String
    /// Tabler icon name (e.g. "ti-receipt", "ti-eye"). Stored as a
    /// string so the icon palette can grow without a migration.
    /// NULL → default sparkle icon.
    public var icon: String?
    /// Hex color for the icon background tint. NULL → accent-tint
    /// default.
    public var color: String?
    /// AND-of-clauses array as canonical JSON. See
    /// `SmartCollectionRule.decode` for the parser; the grammar
    /// itself is in the brief §3.
    public var ruleJson: String
    public var position: Int
    /// True for shipped seed collections ("Needs review" today).
    /// Lets us silently update the seeded ruleJson in a later
    /// release without overwriting user-created collections.
    public var isSeeded: Bool
    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public init(
        id: String,
        userId: String,
        name: String,
        icon: String? = nil,
        color: String? = nil,
        ruleJson: String,
        position: Int = 0,
        isSeeded: Bool = false,
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id          = id
        self.userId      = userId
        self.name        = name
        self.icon        = icon
        self.color       = color
        self.ruleJson    = ruleJson
        self.position    = position
        self.isSeeded    = isSeeded
        self.driveFileId = driveFileId
        self.createdAt   = createdAt
        self.updatedAt   = updatedAt
        self.dirty       = dirty
        self.deletedAt   = deletedAt
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case userId      = "user_id"
        case name
        case icon
        case color
        case ruleJson    = "rule_json"
        case position
        case isSeeded    = "is_seeded"
        case driveFileId = "drive_file_id"
        case createdAt   = "created_at"
        case updatedAt   = "updated_at"
        case dirty
        case deletedAt   = "deleted_at"
    }
}
